### PR TITLE
FF + Prop merging

### PR DESCRIPTION
**Pull request by jacksontromero**
Modifies lora-specific code to now support generic "adapters" that are LoRA and/or Propulsion - propulsion on its own has been tested and is working - it should work w/ LoRA and Propulsion at the same time but that hasn't been thoroughly tested

Note - PR is massive, ignore log files and everything in the `custom_qwen` folder is copied boilerplate from torchtune source code with a ton of relatively small one-line changes from their default versions - don't dive deep here unless something in this code starts breaking

### CODE REVIEW

File: tuning/adapter_tune/README.md
Comment (jacksontromero) [inline tuning/adapter_tune/README.md:1]: @aalexchengg @lizchu413 Overview of adapter tuning code here

File: tuning/adapter_tune/debug_prop_config.yaml [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  # ff_collate_fn: None
+  ff_dataset:
+    _component_: tuning.custom_datasets.codexglue.text_completion_dataset
+    source: google/code_x_glue_cc_code_completion_token
+    column: code
+    split: test[:32]
+    verbose: False # Enable verbose logging in the dataset
+    packed: True # VERY IMPORTANT - Uncommented to ensure proper packing is used
+
+seed: 11868
```
Comment (aalexchengg) [inline tuning/adapter_tune/debug_prop_config.yaml]: seed should be 15122

File: tuning/adapter_tune/debug_lora_prop_config.yaml [hunk 2]
```diff
@@ -62,25 +71,47 @@ resume_from_checkpoint: False
...
+
+  # TODO - CHANGE THIS TO BE A 'REAL' VALIDATION SET
+  # ff_collate_fn: None
+  ff_dataset:
+    _component_: tuning.custom_datasets.codexglue.text_completion_dataset
+    source: google/code_x_glue_cc_code_completion_token
+    column: code
+    split: test[:32]
+    verbose: False # Enable verbose logging in the dataset
+    packed: True # VERY IMPORTANT - Uncommented to ensure proper packing is used
```
Comment (aalexchengg) [inline tuning/adapter_tune/debug_lora_prop_config.yaml:103]: what is packing?

File: tuning/adapter_tune/debug_lora_prop_config.yaml [hunk 3]
```diff
@@ -62,25 +71,47 @@ resume_from_checkpoint: False
...
-batch_size: 5
+batch_size: 1
 
 # Optimizer and Scheduler
 optimizer:
   _component_: torch.optim.AdamW
   fused: True
   weight_decay: 0.01
-  lr: 1e-4
+  lr: 2e-3
```
Comment (aalexchengg) [inline tuning/adapter_tune/debug_lora_prop_config.yaml]: some documentation on why this change was made would be nice

File: tuning/adapter_tune/debug_prop_config.yaml [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+shuffle: True
+batch_size: 1
```
Comment (aalexchengg) [inline tuning/adapter_tune/debug_prop_config.yaml:107]: is there a reason why batch size is so low?

File: tuning/adapter_tune/logs/qwen2_5_1_5b/lora_single_device_ff/log_1744779040.txt
Comment (aalexchengg) [inline tuning/adapter_tune/logs/qwen2_5_1_5b/lora_single_device_ff/log_1744779040.txt:1]: empty file, remove from push if possible

File: tuning/custom_qwen/peft/__init__.py
Comment (aalexchengg) [inline tuning/custom_qwen/peft/__init__.py:1]: I have no clue what is happening here

File: tuning/custom_qwen/qwen2_5/convert_weights.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    except KeyError as e:
+        raise Exception(
+            f'Error converting the state dict. Found unexpected key: "{key}". '
+            "Please make sure you're loading a checkpoint with the right format. "
+        ) from e
+
+    return new_key
+
+
+# def meta_to_tune(state_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
```
Comment (aalexchengg) [inline tuning/custom_qwen/qwen2_5/convert_weights.py:73]: dead code

File: tuning/adapter_tune/debug_lora_prop_config.yaml (hunk 2 above)
Comment (jacksontromero) [reply tuning/adapter_tune/debug_lora_prop_config.yaml:103]: Pre-processes samples/batches into larger chunks up to max seq len, improves training speed a lot by utilizing full GPU https://pytorch.org/torchtune/stable/basics/packing.html

File: tuning/adapter_tune/debug_prop_config.yaml (hunk 4 above)
Comment (jacksontromero) [reply tuning/adapter_tune/debug_prop_config.yaml:107]: This is the debug config, the idea being you can use it to just do a full iter for dev/testing, the main config is `prop_config.yaml`

File: tuning/custom_qwen/qwen2_5/convert_weights.py (hunk 5 above)
Comment (jacksontromero) [reply tuning/custom_qwen/qwen2_5/convert_weights.py:73]: This is a function that came from the base torchtune `convert_weights.py` - For now I think we leave it commented out just to know "hey we can make our own version of this if we have to" but for now I don't think we have to cause we're not using Meta-spec models

File: tuning/custom_recipes/fim/fim_adapter_recipe.py [hunk 6]
```diff
@@ -786,6 +960,99 @@ def train(self) -> None:
                         num_tokens = 0
                         t0 = time.perf_counter()
 
+                    if (
+                        self.do_ff
```
Comment (aalexchengg) [inline tuning/custom_recipes/fim/fim_adapter_recipe.py:967]: inline documentation for major steps would be great here.

File: tuning/adapter_tune/debug_lora_config.yaml [hunk 7]
```diff
@@ -17,8 +17,12 @@
 
 output_dir: ./torchtune/qwen2_5_0_5b/debug_lora_single_device # /tmp may be deleted by your system. Change it to your preference.
```
Comment (aalexchengg) [inline tuning/adapter_tune/debug_lora_config.yaml:18]: why isnt this also changed?

**Comment by aalexchengg** [comment]
might also want to consider removing the term "debug" for the modules/scripts that we should be using, or making new versions that aren't the debugging versions. This PR is massive - in the future, might want to break it down into multiple PRs for easier reviewing.

**Comment by aalexchengg** [comment]
otherwise lgtm

**Comment by jacksontromero** [comment]
> might also want to consider removing the term "debug" for the modules/scripts that we should be using, or making new versions that aren't the debugging versions

@aalexchengg If it has "debug" in it, it is for debugging. That's why they have things like batch_size=1, and there are non-debug configs like `prop_config.yaml`

File: tuning/adapter_tune/debug_lora_config.yaml (hunk 7 above)
Comment (jacksontromero) [reply tuning/adapter_tune/debug_lora_config.yaml:18]: Good catch 👍
